Port `eq` kernel to structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60942 Port other comparison ops: `ne`, `lt`, `gt`, `le`, `ge` to structured kernels.
* **#60177 Port `eq` kernel to structured kernels.**

Tracking issue: #55070

Differential Revision: [D29509871](https://our.internmc.facebook.com/intern/diff/D29509871)

ysiraichi committed:
Port `eq` kernel to structured kernels.

Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `eq` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `eq` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `eq` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi commented:
The 3 CI failures don't seem related to this PR.

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29509871).

ezyang commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -135,6 +162,29 @@ TORCH_META_FUNC(fmin) (const Tensor& self, const Tensor& other) {
+
+TORCH_META_FUNC2(eq, Scalar)(const Tensor& self, const Scalar& other) {
+  auto other_tensor = native::wrapped_scalar_tensor_and_check_convert(other, self);
+  comparison_op_check(self, other_tensor);
```
with wrapped_scalar_tensor_and_check_convert do you really need comparison_op_check?

bdhirsh replied:
huh - if `self` is Scalar, won't that mean that both `self` and `other` have `dim() == 0`, which will make `comparison_op_check` a noop? Since it only does the check if exactly one of the two arguments has zero dimension.

ezyang replied:
that might be true

ysiraichi replied:
That makes sense. I will remove the call

ezyang commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -1020,6 +1054,16 @@ Tensor& greater_equal_out(const Tensor& self, const Scalar& other, Tensor& resul
+(const Tensor& self, const Scalar& other, const Tensor& result) {
+  eq_stub(device_type(), *this);
+}
+
```
Delete the lines below?

ysiraichi replied:
Oops. I think I amended to the wrong commit.

ezyang commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -135,6 +162,29 @@ TORCH_META_FUNC(fmin) (const Tensor& self, const Tensor& other) {
+      native::check_convert(other.item(), self.scalar_type());
+    } else if (self.dim() == 0 && other.dim() != 0) {
+      native::check_convert(self.item(), other.scalar_type());
+    }
```
Note to reviewers: this is ported code that cannot be deleted (and is not otherwise modified); hopefully to be deleted next PR

ezyang replied:
@ngimel this logic looks potentially inefficient: if I have a CUDA scalar and I pass it to a comparison operation, I'm gonna sync for not very good reason.

bdhirsh replied:
why will be be able to remove it in a future PR?

ezyang replied:
no, as currently specced this is impossible to remove unless you turn it into an async cuda assert

ngimel replied:
FYI, now people are rightfully complaining about it (and sorry I missed the original ping a year ago, lol) D36012582

ysiraichi committed:
Update on "Port `eq` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29509871](https://our.internmc.facebook.com/intern/diff/D29509871)

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -799,6 +799,35 @@ void TensorIteratorBase::build_borrowing_binary_float_op(const Tensor& out, cons
+  config.allow_cpu_scalars(true);
+  config.promote_inputs_to_common_dtype(true);
+
+  if (!out.defined()) {
```
Might be worth a comment here - when the output isn't defined yet, e.g. for functional ops like `a == b`, we want the output to be `bool`. When it is defined, e.g. in `torch.eq(a, b, out=c)`, we don't coerce the output type.

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29509871).

ysiraichi committed:
Update on "Port `eq` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29509871](https://our.internmc.facebook.com/intern/diff/D29509871)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `eq` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29509871](https://our.internmc.facebook.com/intern/diff/D29509871)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29509871).

ysiraichi commented:
The CI failures don't seem to be related to this PR.

ezyang committed:
Update on "Port `eq` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29509871](https://our.internmc.facebook.com/intern/diff/D29509871)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29509871).
